Docker Image Resource

jonathangold opened the pull request:
Would you be opposed to moving `examples/lang/docker_container0.mcl` and `examples/lang/docker_image0.mcl` into a folder, so I could import them in `examples/lang/docker0.mcl` instead of duplicating the code?

jonathangold commented:
@purpleidea missed a comment. Fixed and re-pushed.

purpleidea commented on engine/resources/docker_image.go:
```diff
@@ -0,0 +1,267 @@
+			}
+			send = true
+			obj.init.Dirty() // dirty
+		case err, ok := <-errChan:
```
Looks like we need to handle io.EOF specially... Did I read this correctly?

> the stream has been completely read an io.EOF error will be sent over the error channel.

https://github.com/moby/moby/blob/e6fe7f8f29364345083d466a3dfb21f9d539eda0/client/events.go#L15

jonathangold replied:
It is only completely read when we cancel the context, after we return. It doesn't deadlock because they use a buffered channel (see below.)

purpleidea commented on engine/resources/docker_image.go:
```diff
@@ -0,0 +1,267 @@
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *DockerImageRes) Watch() error {
+	ctx, cancel := context.WithCancel(context.Background())
```
What's most troubling is that when we send defer() this looks like it will deadlock:

https://github.com/moby/moby/blob/e6fe7f8f29364345083d466a3dfb21f9d539eda0/client/events.go#L48

```
case <-ctx.Done():
		errs <- ctx.Err()	# wat!
		return
```

I think this is just totally crap on their part, and you should probably report it upstream, but what it means is that we need to "drain" the errChan after close... Which means we need a waitgroup in Watch(), and right after the close happens, we need a defer that has a wg and that starts a go routine to drain the errChan.

```
defer func() { # docker API is WAT.
	wg.Add(1)
	go func() {
		defer wg.Done()
		for {
			_, ok := <- errChan
			if !ok {
				return
			}
		}
	}()
}()
```

Did this not deadlock? Am I reading this all correctly?

jonathangold replied:
It doesn't deadlock, because it's a buffered channel: 

https://github.com/moby/moby/blob/e6fe7f8f29364345083d466a3dfb21f9d539eda0/client/events.go#L22

It's a crappy pattern but it's not a blocking issue, afaict.

purpleidea commented on engine/resources/docker_image.go:
```diff
@@ -0,0 +1,267 @@
+	traits.Edgeable
+
+	// State of the container must be exists or absent.
+	State string `yaml:"state"`
```
So if we want "nginx:latest" how do we do that? Or if we want a specific version? All in the name? (If so, perhaps you could add a small doc note in the struct main docstring about this please.

jonathangold replied:
`image:tag` works, but in order to avoid collisions, I think it makes sense to append the `latest` tag if none is specified. E.g. `nginx` becomes `nginx:latest`.

purpleidea commented on examples/lang/docker_image0.mcl:
```diff
@@ -0,0 +1,3 @@
+docker:image "nginx" {
+    state => "exists",
```
tabs not spaces

jonathangold replied:
:P

purpleidea commented on examples/lang/docker0.mcl:
```diff
@@ -0,0 +1,10 @@
+}
+
+docker:image "nginx" {
+    state => "exists",
```
tabs!

jonathangold replied:
Too much yaml in my life

purpleidea commented:
Thanks for the patches! I didn't understand your question about the examples.

jonathangold commented:
Will fix and rebase this weekend. 

See comment about tags above. In that case `nginx` and `nginx:latest` would be the same, but `nginx:1.14-alpine` would not. 

Vis-a-vis the examples. If I make a folder: `examples/lang/docker/`, I could have a docker example in the `examples/lang` that could import resources from there. It's totally unnecessary right now, but could be nice if and when there are more docker resources.

purpleidea commented:
@jonathangold Heya, not sure if this stalled on my side or yours, but if it's ready for review, could you rebase to master and LMK? Thanks!

jonathangold commented:
@purpleidea I rebased and fixed Watch() to work with the current engine implementation. I haven't looked at the code in quite a while, but it still works as expected. Pretty sure I addressed all your comments.

purpleidea commented:
Sorry for the delay in re-reviewing this. I was really sick for a while, and busy a bit :( Anyways, I decided to merge! \o/ Thanks again! There were a few small nitpicks which I added in a new commit after, most line length stuff (I wish we had a test for this!) and some small Cmp bugs.

Thanks again!
